File: melspec/melspec.go
// Package melspec computes log-Mel-spectrogram features in the
// torchaudio / NeMo convention, distinct from the Kaldi convention
// implemented in github.com/accretional/vad/fbank.
//
// The two packages overlap conceptually — both convert audio to log-Mel
// frames — but the details diverge in ways that matter for model parity:
//
//	             fbank (Kaldi style)         melspec (NeMo/torchaudio style)
//	window       Povey or Hamming (L-1)      Hann periodic=True (L)
//	stft         snip_edges=true             center=True, reflect padding
//	preemph      per-frame (after framing)   whole signal (before framing)
//	mel scale    HTK (1127·ln(1+f/700))      Slaney piecewise (norm=slaney)
//	fmin         20 Hz typical               0 Hz
//	log          log(max(mel, ε))            log(mel + ε), ε = 2^-24
//
// Using the wrong convention against a model produces "numerically plausible
// but model-broken" features — VAD treats real audio as noise. Parity vs
// torchaudio.transforms.MelSpectrogram is verified in melspec_test.go.
package melspec

import (
	"fmt"
	"math"
)

// Options configure a MelSpec instance. NeMoDefaults() returns the settings
// used by NeMo's AudioToMelSpectrogramPreprocessor for the MarbleNet VAD
// model; override fields when wiring up other models.
type Options struct {
	SampleRate    int
	WinLenSamples int     // window length in samples (typical 400 @ 16 kHz)
	HopLenSamples int     // STFT hop in samples (typical 160 @ 16 kHz)
	NFFT          int     // FFT size; must be >= WinLenSamples and power of 2
	NumMelBins    int     // typical 80
	FMinHz        float64 // typical 0 (NeMo) or 20 (Kaldi)
	FMaxHz        float64 // 0 means SampleRate/2
	PreemphCoeff  float64 // 0.97 typical; set to 0 to disable
	LogOffset     float64 // 2^-24 typical (NeMo log_zero_guard_value); set to 0 for no offset
	PadToMultiple int     // right-pad time axis with zeros to this multiple; 0 disables
}

// NeMoDefaults returns the exact configuration used by NVIDIA NeMo's
// AudioToMelSpectrogramPreprocessor for the MarbleNet VAD model at 16 kHz.
// These values match the preprocessor.yaml saved alongside the ONNX export.
func NeMoDefaults() Options {
	return Options{
		SampleRate:    16000,
		WinLenSamples: 400, // 25 ms
		HopLenSamples: 160, // 10 ms
		NFFT:          512,
		NumMelBins:    80,
		FMinHz:        0,
		FMaxHz:        0, // ⇒ 8000
		PreemphCoeff:  0.97,
		LogOffset:     math.Ldexp(1, -24), // 2^-24
		PadToMultiple: 2,
	}
}

// MelSpec computes log-Mel features. Construct once (window + filterbank are
// precomputed); Compute is safe to call repeatedly. Not safe for concurrent
// calls on the same instance.
type MelSpec struct {
	opts    Options
	window  []float64 // Hann window of length WinLenSamples
	mel     [][]float64 // [NumMelBins][fftBins] Slaney filterbank
	fftBins int       // NFFT/2 + 1

	// Per-call scratch space reused across frames.
	frame  []float64 // length NFFT
	fftBuf []complex128
}

// New constructs a MelSpec with the given options.
func New(opts Options) (*MelSpec, error) {
	if opts.SampleRate <= 0 || opts.WinLenSamples <= 0 || opts.HopLenSamples <= 0 {
		return nil, fmt.Errorf("melspec: SampleRate, WinLenSamples, HopLenSamples must be > 0")
	}
	if opts.NumMelBins <= 0 {
		return nil, fmt.Errorf("melspec: NumMelBins must be > 0")
	}
	if opts.NFFT < opts.WinLenSamples || (opts.NFFT&(opts.NFFT-1)) != 0 {
		return nil, fmt.Errorf("melspec: NFFT must be a power of 2 >= WinLenSamples")
	}
	fmax := opts.FMaxHz
	if fmax <= 0 {
		fmax = float64(opts.SampleRate) / 2
	}
	fftBins := opts.NFFT/2 + 1
	return &MelSpec{
		opts:    opts,
		window:  hannPeriodic(opts.WinLenSamples),
		mel:     makeMelFilterbankSlaney(opts.NumMelBins, fftBins, float64(opts.SampleRate), opts.FMinHz, fmax),
		fftBins: fftBins,
		frame:   make([]float64, opts.NFFT),
		fftBuf:  make([]complex128, opts.NFFT),
	}, nil
}

// Compute returns log-mel-spectrogram features for the given mono float32
// PCM samples in [-1, 1] (the same convention models expect at the API
// boundary). Output shape is `[T_mel][NumMelBins]` — row-major, time-first.
//
// To pass to a NeMo-style ONNX model expecting `[1, NumMelBins, T_mel]`
// (channels-first), the caller transposes — see ComputeChannelsFirstFlat.
func (m *MelSpec) Compute(samples []float32) [][]float32 {
	tMel, padded := m.computeImpl(samples)
	if tMel == 0 {
		return nil
	}
	out := make([][]float32, tMel)
	for t := 0; t < tMel; t++ {
		out[t] = padded[t]
	}
	return out
}

// ComputeChannelsFirstFlat is a convenience wrapper that returns the flat
// `[1, NumMelBins, T_mel]` tensor ready to feed straight to a NeMo ONNX
// session. Returns (flat, tMel). T_mel matches the time dim after pad_to.
func (m *MelSpec) ComputeChannelsFirstFlat(samples []float32) ([]float32, int) {
	tMel, frames := m.computeImpl(samples)
	if tMel == 0 {
		return nil, 0
	}
	flat := make([]float32, m.opts.NumMelBins*tMel)
	for t := 0; t < tMel; t++ {
		row := frames[t]
		for b := 0; b < m.opts.NumMelBins; b++ {
			flat[b*tMel+t] = row[b]
		}
	}
	return flat, tMel
}

// computeImpl runs the full pipeline: preemphasis → reflect-pad → STFT →
// power → mel filterbank → log → pad_to. Returns (T_mel, frames[T_mel][NumMelBins]).
func (m *MelSpec) computeImpl(samples []float32) (int, [][]float32) {
	n := len(samples)
	if n == 0 {
		return 0, nil
	}

	// 1. Pre-emphasize the whole signal once. NeMo's exact formula:
	//    y[0] = x[0]; y[i] = x[i] - preemph * x[i-1] for i >= 1.
	preemphed := make([]float64, n)
	preemphed[0] = float64(samples[0])
	if m.opts.PreemphCoeff != 0 {
		for i := 1; i < n; i++ {
			preemphed[i] = float64(samples[i]) - m.opts.PreemphCoeff*float64(samples[i-1])
		}
	} else {
		for i := 1; i < n; i++ {
			preemphed[i] = float64(samples[i])
		}
	}

	// 2. Reflection pad by NFFT/2 on each side (matches torch.stft center=True
	//    with pad_mode='reflect').
	pad := m.opts.NFFT / 2
	padded := make([]float64, n+2*pad)
	for i := 0; i < pad; i++ {
		// Reflect: padded[i] = preemphed[pad - i] (mirror around index 0).
		padded[i] = preemphed[pad-i]
	}
	copy(padded[pad:pad+n], preemphed)
	for i := 0; i < pad; i++ {
		// Reflect at the right edge: padded[pad+n+i] = preemphed[n-2-i].
		idx := n - 2 - i
		if idx < 0 {
			idx = 0
		}
		padded[pad+n+i] = preemphed[idx]
	}

	// 3. Frame: each frame starts at i*hop in the padded signal.
	//    NeMo's get_seq_len is floor((N + 2*pad - n_fft) / hop) — no +1.
	//    torch.stft natively produces ONE MORE frame than this; NeMo masks
	//    that trailing frame to 0 before pad_to. We match by producing only
	//    NeMo's count up front (omitting the extra trailing frame).
	tMel := (len(padded) - m.opts.NFFT) / m.opts.HopLenSamples
	if tMel <= 0 {
		return 0, nil
	}

	// Compute padded T_mel (after pad_to right-padding with zeros).
	finalT := tMel
	if m.opts.PadToMultiple > 0 {
		rem := finalT % m.opts.PadToMultiple
		if rem != 0 {
			finalT += m.opts.PadToMultiple - rem
		}
	}

	out := make([][]float32, finalT)
	for t := 0; t < tMel; t++ {
		out[t] = make([]float32, m.opts.NumMelBins)
		m.computeFrame(padded[t*m.opts.HopLenSamples:t*m.opts.HopLenSamples+m.opts.NFFT], out[t])
	}
	// pad_to=N right-pads the time axis with LITERAL zeros (matching torch's
	// nn.functional.pad default mode='constant' value=0 — applied to the
	// already-logged feature matrix, NOT log-of-zero). The padded frames
	// look like log-energy = 0 (=> unit linear), which is technically
	// "louder than silence" but the model is trained to handle these padding
	// frames as such.
	for t := tMel; t < finalT; t++ {
		out[t] = make([]float32, m.opts.NumMelBins)
	}
	return finalT, out
}

// computeFrame applies window + FFT + power + mel + log to one frame's worth
// of samples (length must be NFFT). When WinLenSamples < NFFT, the window
// is CENTERED inside the NFFT buffer with (NFFT-WinLenSamples)/2 zeros on
// each side — matching torch.stft's behaviour. Getting this wrong silently
// produces "windows over the wrong 400 samples per frame" which yields a
// numerically-plausible but model-incompatible spectrogram.
func (m *MelSpec) computeFrame(in []float64, out []float32) {
	for i := 0; i < m.opts.NFFT; i++ {
		m.frame[i] = 0
	}
	winOffset := (m.opts.NFFT - m.opts.WinLenSamples) / 2
	for i := 0; i < m.opts.WinLenSamples; i++ {
		m.frame[winOffset+i] = in[winOffset+i] * m.window[i]
	}
	// FFT: real input → complex spectrum.
	for i := 0; i < m.opts.NFFT; i++ {
		m.fftBuf[i] = complex(m.frame[i], 0)
	}
	fftRadix2(m.fftBuf)
	// Power spectrum |X[k]|^2 for k = 0..fftBins.
	power := make([]float64, m.fftBins)
	for k := 0; k < m.fftBins; k++ {
		re := real(m.fftBuf[k])
		im := imag(m.fftBuf[k])
		power[k] = re*re + im*im
	}
	// Slaney mel filterbank * power, then log(mel + log_offset).
	for b := 0; b < m.opts.NumMelBins; b++ {
		row := m.mel[b]
		var sum float64
		for k := 0; k < m.fftBins; k++ {
			sum += row[k] * power[k]
		}
		out[b] = float32(math.Log(sum + m.opts.LogOffset))
	}
}

// NumFrames returns how many output frames Compute will produce for the
// given input length (after pad_to). Useful for sizing buffers.
func (m *MelSpec) NumFrames(numSamples int) int {
	if numSamples <= 0 {
		return 0
	}
	pad := m.opts.NFFT / 2
	t := (numSamples + 2*pad - m.opts.NFFT) / m.opts.HopLenSamples
	if t <= 0 {
		return 0
	}
	if m.opts.PadToMultiple > 0 {
		rem := t % m.opts.PadToMultiple
		if rem != 0 {
			t += m.opts.PadToMultiple - rem
		}
	}
	return t
}

// hannPeriodic returns a Hann window of length L using the "periodic"
// convention: w[i] = 0.5 - 0.5*cos(2π*i/L). torchaudio's hann_window with
// periodic=True (the default).
func hannPeriodic(length int) []float64 {
	w := make([]float64, length)
	if length == 0 {
		return w
	}
	denom := float64(length)
	for i := 0; i < length; i++ {
		w[i] = 0.5 - 0.5*math.Cos(2*math.Pi*float64(i)/denom)
	}
	return w
}

File: cmd/basic-vad-web/static/js/dsp/melspec.js
// Port of melspec/melspec.go + melspec/mel.go.
//
// NeMo / torchaudio MelSpectrogram: Hann periodic window, Slaney mel scale,
// STFT center=True with reflect padding, whole-signal preemph, log(mel + 2^-24).
// Used by the MarbleNet VAD backend.

import { fftRadix2 } from './fft.js';

export function nemoDefaults() {
    return {
        sampleRate: 16000,
        winLenSamples: 400,   // 25 ms
        hopLenSamples: 160,   // 10 ms
        nFFT: 512,
        numMelBins: 80,
        fMinHz: 0,
        fMaxHz: 0,            // 0 ⇒ sampleRate/2
        preemphCoeff: 0.97,
        logOffset: Math.pow(2, -24),
        padToMultiple: 2,
    };
}

// Slaney piecewise mel scale (librosa htk=False).
const SLANEY_F_SP = 200.0 / 3.0;
const SLANEY_MIN_LOG_HZ = 1000.0;
const SLANEY_MIN_LOG_MEL = SLANEY_MIN_LOG_HZ / SLANEY_F_SP;
const SLANEY_LOG_STEP = Math.log(6.4) / 27.0;

function hzToMelSlaney(hz) {
    if (hz < SLANEY_MIN_LOG_HZ) return hz / SLANEY_F_SP;
    return SLANEY_MIN_LOG_MEL + Math.log(hz / SLANEY_MIN_LOG_HZ) / SLANEY_LOG_STEP;
}
function melToHzSlaney(mel) {
    if (mel < SLANEY_MIN_LOG_MEL) return mel * SLANEY_F_SP;
    return SLANEY_MIN_LOG_HZ * Math.exp(SLANEY_LOG_STEP * (mel - SLANEY_MIN_LOG_MEL));
}

function makeMelFilterbankSlaney(numBins, fftBins, sampleRate, lowFreq, highFreq) {
    const fftSize = (fftBins - 1) * 2;
    if (highFreq <= 0 || highFreq > sampleRate / 2) highFreq = sampleRate / 2;
    const melLow = hzToMelSlaney(lowFreq);
    const melHigh = hzToMelSlaney(highFreq);
    const melPoints = new Float64Array(numBins + 2);
    for (let i = 0; i < melPoints.length; i++) {
        melPoints[i] = melLow + (melHigh - melLow) * i / (numBins + 1);
    }
    const hzPoints = new Float64Array(melPoints.length);
    for (let i = 0; i < melPoints.length; i++) hzPoints[i] = melToHzSlaney(melPoints[i]);
    const binHz = new Float64Array(fftBins);
    for (let k = 0; k < fftBins; k++) binHz[k] = k * sampleRate / fftSize;
    const filt = [];
    for (let b = 0; b < numBins; b++) {
        const left = hzPoints[b], center = hzPoints[b + 1], right = hzPoints[b + 2];
        // Slaney area normalization: enorm = 2 / (right - left).
        const enorm = 2.0 / (right - left);
        const row = new Float64Array(fftBins);
        for (let k = 0; k < fftBins; k++) {
            const f = binHz[k];
            if (f < left || f > right) continue;
            if (f <= center) row[k] = (f - left) / (center - left) * enorm;
            else            row[k] = (right - f) / (right - center) * enorm;
        }
        filt.push(row);
    }
    return filt;
}

// Hann periodic: w[i] = 0.5 - 0.5*cos(2π*i/L). torchaudio default.
function hannPeriodic(length) {
    const w = new Float64Array(length);
    for (let i = 0; i < length; i++) {
        w[i] = 0.5 - 0.5 * Math.cos(2 * Math.PI * i / length);
    }
    return w;
}

export class MelSpec {
    constructor(opts = nemoDefaults()) {
        this.opts = { ...nemoDefaults(), ...opts };
        const o = this.opts;
        if (o.nFFT < o.winLenSamples || (o.nFFT & (o.nFFT - 1)) !== 0) {
            throw new Error('melspec: nFFT must be a power of 2 >= winLenSamples');
        }
        this.fftBins = (o.nFFT >> 1) + 1;
        const fmax = o.fMaxHz > 0 ? o.fMaxHz : o.sampleRate / 2;
        this.window = hannPeriodic(o.winLenSamples);
        this.mel = makeMelFilterbankSlaney(o.numMelBins, this.fftBins, o.sampleRate, o.fMinHz, fmax);
        this._re = new Float64Array(o.nFFT);
        this._im = new Float64Array(o.nFFT);
        this._power = new Float64Array(this.fftBins);
    }

    // computeChannelsFirstFlat returns { feats, tMel } where feats is a
    // Float32Array of length numMelBins * tMel, laid out as [b * tMel + t]
    // (matching Go's ComputeChannelsFirstFlat). Ready to feed a NeMo ONNX
    // model with input shape [1, numMelBins, tMel].
    computeChannelsFirstFlat(samples) {
        const o = this.opts;
        const n = samples.length;
        if (n === 0) return { feats: null, tMel: 0 };

        // 1. Whole-signal preemphasis.
        const preemphed = new Float64Array(n);
        preemphed[0] = samples[0];
        if (o.preemphCoeff !== 0) {
            for (let i = 1; i < n; i++) {
                preemphed[i] = samples[i] - o.preemphCoeff * samples[i - 1];
            }
        } else {
            for (let i = 1; i < n; i++) preemphed[i] = samples[i];
        }

        // 2. Reflection pad by nFFT/2 on each side (torch.stft pad_mode=reflect).
        const pad = o.nFFT >> 1;
        const padded = new Float64Array(n + 2 * pad);
        for (let i = 0; i < pad; i++) padded[i] = preemphed[pad - i];
        padded.set(preemphed, pad);
        for (let i = 0; i < pad; i++) {
            let idx = n - 2 - i;
            if (idx < 0) idx = 0;
            padded[pad + n + i] = preemphed[idx];
        }

        // 3. Frame count (matches NeMo get_seq_len: no +1 vs torch.stft).
        const tMel = Math.floor((padded.length - o.nFFT) / o.hopLenSamples);
        if (tMel <= 0) return { feats: null, tMel: 0 };
        let finalT = tMel;
        if (o.padToMultiple > 0) {
            const rem = finalT % o.padToMultiple;
            if (rem !== 0) finalT += o.padToMultiple - rem;
        }

        const nBins = o.numMelBins;
        const feats = new Float32Array(nBins * finalT);
        const winOffset = (o.nFFT - o.winLenSamples) >> 1;
        for (let t = 0; t < tMel; t++) {
            this._computeFrame(padded, t * o.hopLenSamples, winOffset, feats, t, finalT);
        }
        // pad_to right-pads with literal zeros (log-of-1 region); leave as 0.
        return { feats, tMel: finalT };
    }

    _computeFrame(padded, srcOff, winOffset, out, t, tMel) {
        const o = this.opts;
        const re = this._re;
        const im = this._im;
        for (let i = 0; i < o.nFFT; i++) { re[i] = 0; im[i] = 0; }
        for (let i = 0; i < o.winLenSamples; i++) {
            re[winOffset + i] = padded[srcOff + winOffset + i] * this.window[i];
        }
        fftRadix2(re, im);
        const fftBins = this.fftBins;
        const power = this._power;
        for (let k = 0; k < fftBins; k++) power[k] = re[k] * re[k] + im[k] * im[k];
        const nBins = o.numMelBins;
        const offs = o.logOffset;
        for (let b = 0; b < nBins; b++) {
            const row = this.mel[b];
            let sum = 0;
            for (let k = 0; k < fftBins; k++) sum += row[k] * power[k];
            out[b * tMel + t] = Math.log(sum + offs);
        }
    }
}
